sdk/src/trace/batch_span_processor.cc

```diff
@@ -0,0 +1,184 @@
+#pragma once
+
+#include "opentelemetry/sdk/trace/batch_span_processor.h"
+#include "src/common/circular_buffer.h"
+
+#include <vector>
+
+using opentelemetry::sdk::common::AtomicUniquePtr;
+using opentelemetry::sdk::common::CircularBuffer;
+using opentelemetry::sdk::common::CircularBufferRange;
+
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace trace
+{   
+
+BatchSpanProcessor::BatchSpanProcessor(std::unique_ptr<SpanExporter>&& exporter,
+                                       const int max_queue_size,
+                                       const int schedule_delay_millis,
+                                       const int max_export_batch_size)
+       :exporter_(std::move(exporter)),
+        max_queue_size_(max_queue_size), 
+        schedule_delay_millis_(schedule_delay_millis),
+        max_export_batch_size_(max_export_batch_size)   
+{  
+    buffer_ = std::unique_ptr<CircularBuffer<Recordable>>(new CircularBuffer<Recordable>(max_queue_size_));
+    
+    //Start the background worker thread
+    worker_thread_ = std::unique_ptr<std::thread>(new std::thread(&BatchSpanProcessor::DoBackgroundWork, this));
+
+} 
+
+std::unique_ptr<Recordable> BatchSpanProcessor::MakeRecordable() noexcept 
+{
+    return exporter_->MakeRecordable();
+}
+
+void BatchSpanProcessor::OnStart(Recordable &span) noexcept  
+{
+    // no-op
+}
+
+void BatchSpanProcessor::OnEnd(std::unique_ptr<Recordable> &&span) noexcept 
+{
+    if(is_shutdown_){
+        // TODO: log in glog
+        return;
+    }
+
+    if(static_cast<int>(buffer_->size()) == max_queue_size_){
+        // TODO: glog that spans will likely be dropped
+    }
+
+    buffer_->Add(span);
+
+    // If the queue gets half full a preemptive notification is sent to the worker thread to
+    // start a new export cycle.
+    if(static_cast<int>(buffer_->size()) == max_queue_size_ / 2){
+        // signal the worker thread
+        cv_.notify_one();
+    }
+}
+
+void BatchSpanProcessor::ForceFlush(std::chrono::microseconds timeout) noexcept 
+{    
+    if(is_shutdown_ == true){
+        // TODO: glog that processor is already shutdown
+    }
+
+    is_force_flush_ = true;
+    buffer_->Add(FORCE_FLUSH_TOKEN);
+
+    // wake up the worker thread
+    cv_.notify_one();
+
+    // wait on flush token span to be processed
+    std::unique_lock<std::mutex> force_flush_lk(force_flush_cv_m_);
+    auto status = force_flush_cv_.wait_for(force_flush_lk, timeout);
```

Comment:
Over here, spurious wake-ups are not handled, and the same goes for [python](https://github.com/open-telemetry/opentelemetry-python/blob/master/opentelemetry-sdk/src/opentelemetry/sdk/trace/export/__init__.py#L245).